### PR TITLE
Break after a multiline argument in an argument list

### PR DESCRIPTION
**Pull request by gpetiot**
Found while working on #1352

### CODE REVIEW

File: test/passing/comments_args.ml.ref [hunk 1]
```diff
@@ -5,7 +5,9 @@ let emit_wrapper_function =
     (Hhas_pos.pos_to_span ast_fun.Ast.f_span)
     false (* is_async *) false (* is_generator *) false
     (* is_pair_generator *) hoisted true (* no_injection *) true
-    (* inout_wrapper *) is_interceptable false (* is_memoize_impl *) Rx.NonRx
+    (* inout_wrapper *) is_interceptable false
+    (* is_memoize_impl *)
+    Rx.NonRx
```
Comment (gpetiot) [inline test/passing/comments_args.ml.ref:10]: closer to the input, probably partially due to the comment attachment (will be improved to another PR)

**Comment by gpetiot** [comment]
I don't know why it s breaking in these cases, the multiline check should be enough. But it's an ugly patch mixing the `break_unless_newline` with the `fits_breaks` that were already in the equation, I had low expectations on this one.

**Comment by gpetiot** [comment]
I managed to have something cleaner by only fixing the `Ast.is_simple` function (for exprs). The `:=` changes in the tests look like a fix to me. Can you check on your codebase, and then I will check on other big projects.

File: test/passing/args_grouped.ml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        goooooooo
+  in
+  fooooooooooooooooooooo
+
+let declare_locals_and_ret tenv pdesc (prop_ : Prop.normal Prop.t) =
+  let foooooooooooooo =
+    BiabductionConfig.run_in_re_execution_mode
+      (* no footprint vars for locals *)
+      sigma_locals_and_ret
+      ()
```
Comment (gpetiot) [inline test/passing/args_grouped.ml:40]: the comment is only attached to `sigma_locals_and_ret` so this is a complex expression, causing a break before `()`

**Comment by gpetiot** [comment]
Even if it's not a big regression I would like to fix it before merging this PR.
But I didn't find the solution yet, do you have many occurrences of this issue? do they all follow the same pattern as this one?

**Comment by jberdine** [comment]
From what I could see, they all seemed to have that pattern, yes. It's hard to be sure though.

There were not a large number of changes. Note that infer is now using the 0.14.1-13-g000da99 (aka master) version of ocamlformat, so testing should be much easier than before.

**Comment by gpetiot** [comment]
I think I fixed it, there was escaped `\n` that didn't make the formatted string to break but lead the heuristic to guess the string was breaking.

**Comment by jberdine** [comment]
I tested the current version and found 2 cases that are maybe interesting:
```diff
modified   infer/src/biabduction/SymExec.ml
@@ -1161,7 +1161,8 @@ let declare_locals_and_ret tenv pdesc (prop_ : Prop.normal Prop.t) =
     in
     BiabductionConfig.run_in_re_execution_mode
       (* no footprint vars for locals *)
-      sigma_locals_and_ret ()
+      sigma_locals_and_ret
+      ()
   in
   let sigma' = prop_.Prop.sigma @ sigma_locals_and_ret in
   let prop' = Prop.normalize tenv (Prop.set prop_ ~sigma:sigma') in
```
Here my understanding is that the comment is attached to the `sigma_locals_and_ret` arg, and it is therefore considered to be a multi-line arg, and so there is a break after. This could go either way. I expect that it would be more understandable to users if the decision to break was not as dependent on internal implementation details like the particulars of comment attachment. Not a big deal though.

The other is:
```diff
modified   infer/src/unit/DifferentialFiltersTests.ml
@@ -23,8 +23,8 @@ let test_file_renamings_from_json =
     match expected_output with
     | Return exp ->
         assert_equal ~pp_diff
-          ~cmp:DifferentialFilters.FileRenamings.VISIBLE_FOR_TESTING_DO_NOT_USE_DIRECTLY.equal exp
-          (test_output test_input)
+          ~cmp:DifferentialFilters.FileRenamings.VISIBLE_FOR_TESTING_DO_NOT_USE_DIRECTLY.equal
+          exp (test_output test_input)
     | Raise exc ->
         assert_raises exc (fun () -> test_output test_input)
   in
```
This one I expect is a case of a bit of noise when getting very close to the margin due to conflicting with Format's max_indent. Note that the previous version reaches column 98 with a margin configured to 100.

Those are not blockers.

**Comment by gpetiot** [comment]
Great, before merging this one I would like to test the combination of #1352, #1360 and #1367; if the output of test_branch looks good I will integrate them in the preview `janestreet` branch to check we are not causing regressions to avoid a rollback later.

**Comment by gpetiot** [comment]
Minor update on the PR, here is now the diff of test_branch:

```diff
diff --git a/infer/src/base/Config.ml b/infer/src/base/Config.ml
index 29ce48283..583da2d45 100644
--- a/infer/src/base/Config.ml
+++ b/infer/src/base/Config.ml
@@ -1596,7 +1596,8 @@ and nullsafe_optimistic_third_party_params_in_non_strict =
          Historically this is because there was no actionable way to change third party annotations.
          Now that we have such a support, this behavior should be reconsidered, provided
          our tooling and error reporting is friendly enough to be smoothly used by developers.
-      *) ~default:true
+      *)
+    ~default:true
     "Nullsafe: in this mode we treat non annotated third party method params as if they were \
      annotated as nullable."
 
diff --git a/src/dune/expander.ml b/src/dune/expander.ml
index 70eae332..2f634af2 100644
--- a/src/dune/expander.ml
+++ b/src/dune/expander.ml
@@ -463,7 +463,8 @@ let gen_with_record_deps ~expand t resolved_forms ~dep_kind =
     expand
       (* we keep the dir constant here to replicate the old behavior of: (chdir
          foo %{exe:bar}). This should lookup ./bar rather than ./foo/bar *)
-      resolved_forms ~dir:t.dir ~dep_kind ~expand_var:t.expand_var
+      resolved_forms
+      ~dir:t.dir ~dep_kind ~expand_var:t.expand_var
   in
   { t with expand_var }
 
diff --git a/src/dune/foreign_rules.ml b/src/dune/foreign_rules.ml
index 4987a6f8..3b250147 100644
--- a/src/dune/foreign_rules.ml
+++ b/src/dune/foreign_rules.ml
@@ -88,8 +88,8 @@ let build_c_file ~sctx ~dir ~expander ~include_flags (loc, src, dst) =
     (let src = Path.build (Foreign.Source.path src) in
      Command.run
      (* We have to execute the rule in the library directory as the .o is
-        produced in the current directory *) ~dir:(Path.build dir)
-       (Ok ctx.ocamlc)
+        produced in the current directory *)
+       ~dir:(Path.build dir) (Ok ctx.ocamlc)
        [ A "-g"
        ; include_flags
        ; Dyn (Build.map c_flags ~f:(fun x -> Command.quote_args "-ccopt" x))
@@ -121,7 +121,8 @@ let build_cxx_file ~sctx ~dir ~expander ~include_flags (loc, src, dst) =
      let c_compiler = Ocaml_config.c_compiler ctx.ocaml_config in
      Command.run
      (* We have to execute the rule in the library directory as the .o is
-        produced in the current directory *) ~dir:(Path.build dir)
+        produced in the current directory *)
+       ~dir:(Path.build dir)
        (Super_context.resolve_program ~loc:None ~dir sctx c_compiler)
        ( [ Command.Args.S [ A "-I"; Path ctx.stdlib_dir ]
          ; include_flags
```

**Comment by jberdine** [comment]
The one in dune/expander.ml looks like the same behavior I saw where an arg with an attached comment gets considered to be nonsimple due to the existence/length of the comment. The other changes are bug fixes.

**Comment by gpetiot** [comment]
> This could go either way. I expect that it would be more understandable to users if the decision to break was not as dependent on internal implementation details like the particulars of comment attachment. Not a big deal though.

I also think it's not a big deal but this may cause unwanted diff the next time users upgrade ocamlformat. But I have no idea how to fix this, so if there is no objection I will merge it soon.

**Comment by jberdine** [comment]
Just wondering, but is leaving out preceding and trailing comments when pre-formatting in is_simple an option?

**Comment by gpetiot** [comment]
> Just wondering, but is leaving out preceding and trailing comments when pre-formatting in is_simple an option?

I tried something in this direction in my last commit. @jberdine can you test on your codebase there is no regression? I'm dropping the comments before, to not prevent wrapping in this case:
```
   (* fooooooooooooo *)
   fooo foooooooooooo fooooooo
```

but not after, so we have a break here, and the parameters are not hidden after the comment:
```
   fooooooo
   (* fooooooooooooooo *)
   fooooooooooo
```

The test_branch diff looks good:
```diff
diff --git a/infer/src/base/Config.ml b/infer/src/base/Config.ml
index bd9252ee2..9dbffe6a1 100644
--- a/infer/src/base/Config.ml
+++ b/infer/src/base/Config.ml
@@ -1646,7 +1646,8 @@ and nullsafe_optimistic_third_party_params_in_non_strict =
          Historically this is because there was no actionable way to change third party annotations.
          Now that we have such a support, this behavior should be reconsidered, provided
          our tooling and error reporting is friendly enough to be smoothly used by developers.
-      *) ~default:true
+      *)
+    ~default:true
     "Nullsafe: in this mode we treat non annotated third party method params \
      as if they were annotated as nullable."
 
diff --git a/src/dune/foreign_rules.ml b/src/dune/foreign_rules.ml
index a3fe2af7..92c14441 100644
--- a/src/dune/foreign_rules.ml
+++ b/src/dune/foreign_rules.ml
@@ -90,8 +90,8 @@ let build_c_file ~sctx ~dir ~expander ~include_flags (loc, src, dst) =
     (let src = Path.build (Foreign.Source.path src) in
      Command.run
      (* We have to execute the rule in the library directory as the .o is
-        produced in the current directory *) ~dir:(Path.build dir)
-       (Ok ctx.ocamlc)
+        produced in the current directory *)
+       ~dir:(Path.build dir) (Ok ctx.ocamlc)
        [
          A "-g";
          include_flags;
@@ -124,7 +124,8 @@ let build_cxx_file ~sctx ~dir ~expander ~include_flags (loc, src, dst) =
      let c_compiler = Ocaml_config.c_compiler ctx.ocaml_config in
      Command.run
      (* We have to execute the rule in the library directory as the .o is
-        produced in the current directory *) ~dir:(Path.build dir)
+        produced in the current directory *)
+       ~dir:(Path.build dir)
        (Super_context.resolve_program ~loc:None ~dir sctx c_compiler)
        ( [
            Command.Args.S [ A "-I"; Path ctx.stdlib_dir ];
```

**Comment by gpetiot** [comment]
Can we merge this one?

**Comment by jberdine** [comment]
Sorry to be slow to test, I plan to, just need a chance to context switch.
